drivers/flash/flash_realtek_rts5912.c

```diff
@@ -0,0 +1,768 @@
+}
+
+static int flash_program_page(const struct device *dev, uint32_t address, const uint8_t *data,
+			      uint32_t size)
+{
+	struct flash_rts5912_dev_data *dev_data = dev->data;
+	struct qspi_cmd *command = &dev_data->command_default;
+	SPIC_ADDRESS_SIZE_t addr_size = SPIC_CFG_ADDR_SIZE_24;
+	int ret = 0;
+	uint32_t offset = 0, chunk = 0, page_size = FLASH_PAGE_SZ;
+
+	ret = flash_write_enable(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	config_command(command, SPI_NOR_CMD_PP, address, addr_size, 0);
+
+	while (size > 0) {
+		offset = address % page_size;
+		chunk = (offset + size < page_size) ? size : (page_size - offset);
+
+		command->address.value = address;
+		ret = spic_write(command, data, (size_t *)&chunk);
+		if (ret < 0) {
+			goto err_exit;
+		}
+
+		data += chunk;
+		address += chunk;
+		size -= chunk;
+
+		flash_wait_till_ready(dev);
+	}
+
+err_exit:
+	flash_write_disable(dev);
+	return ret;
+}
+
+static int flash_normal_read(const struct device *dev, uint8_t rdcmd, uint32_t address,
+			     uint8_t *data, uint32_t size)
+{
+	struct flash_rts5912_dev_data *dev_data = dev->data;
+	struct qspi_cmd *command = &dev_data->command_default;
+	SPIC_ADDRESS_SIZE_t addr_size = SPIC_CFG_ADDR_SIZE_24;
+	int ret;
+
+	uint32_t src_addr = address;
+	uint8_t *dst_idx = data;
+
+	uint32_t remind_size = size;
+	uint32_t block_size = 0x8000ul;
+	uint8_t dummy_count = (rdcmd == SPI_NOR_CMD_READ) ? 0 : 8;
+
+	config_command(command, rdcmd, src_addr, addr_size, dummy_count);
+
+	while (remind_size > 0) {
+		command->address.value = src_addr;
+
+		if (remind_size >= block_size) {
+			ret = spic_read(command, dst_idx, (size_t *)&block_size);
+			src_addr += block_size;
+			remind_size -= block_size;
+			dst_idx += block_size;
+		} else {
+			ret = spic_read(command, dst_idx, (size_t *)&remind_size);
+			dst_idx += remind_size;
+			remind_size = 0;
+		}
+
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int flash_rts5912_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct flash_rts5912_dev_data *data = dev->data;
+	int ret = -EINVAL;
+
+	if ((offset % FLASH_ERASE_BLK_SZ) != 0) {
+		return -EINVAL;
+	}
+
+	if ((len % FLASH_ERASE_BLK_SZ) != 0) {
+		return -EINVAL;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	for (; len > 0; len -= FLASH_ERASE_BLK_SZ) {
+		ret = flash_erase_sector(dev, offset);
+		if (ret < 0) {
+			LOG_ERR("erase @0x%08lx fail", offset);
+		}
+		offset += FLASH_ERASE_BLK_SZ;
+	}
+
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+static int flash_rts5912_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_rts5912_dev_data *dev_data = dev->data;
+	int ret;
+	unsigned int key;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+	key = irq_lock();
+	ret = flash_program_page(dev, offset, data, len);
+	irq_unlock(key);
+	k_sem_give(&dev_data->sem);
+
+	return ret;
+}
+
+static int flash_rts5912_read(const struct device *dev, off_t offset, void *data, size_t len)
```

Comment:
No range check on `offset` and `len`.
This should return 0 and do nothing, preferably, when `len` is 0.